src/liballoc_jemalloc/lib.rs

```diff
@@ -201,27 +201,47 @@ mod contents {
                                             align: usize,
                                             excess: *mut usize,
                                             err: *mut u8) -> *mut u8 {
-        let p = __rde_alloc(size, align, err);
-        if !p.is_null() {
-            *excess = size;
+        let flags = align_to_flags(align);
+        let ptr = mallocx(size as size_t, flags) as *mut u8;
+        if ptr.is_null() {
+            let layout = Layout::from_size_align_unchecked(size, align);
+            ptr::write(err as *mut AllocErr,
+                       AllocErr::Exhausted { request: layout });
+        } else {
+            *excess = nallocx(size as size_t, flags) as usize;
         }
-        return p
+        ptr
     }
 
     #[no_mangle]
     #[linkage = "external"]
     pub unsafe extern fn __rde_realloc_excess(ptr: *mut u8,
-                                              old_size: usize,
+                                              _old_size: usize,
                                               old_align: usize,
                                               new_size: usize,
                                               new_align: usize,
                                               excess: *mut usize,
                                               err: *mut u8) -> *mut u8 {
-        let p = __rde_realloc(ptr, old_size, old_align, new_size, new_align, err);
-        if !p.is_null() {
-            *excess = new_size;
+        if new_align != old_align {
+            ptr::write(
+                err as *mut AllocErr,
+                AllocErr::Unsupported { details: "can't change alignments" },
+            );
+            return 0 as *mut u8
         }
-        return p
+
+        let flags = align_to_flags(new_align);
+        let ptr = rallocx(ptr as *mut c_void, new_size, flags) as *mut u8;
+        if ptr.is_null() {
+            let layout = Layout::from_size_align_unchecked(new_size, new_align);
+            ptr::write(
+                err as *mut AllocErr,
+                AllocErr::Exhausted { request: layout },
+            );
+        } else {
+            *excess = nallocx(new_size, flags) as usize;
```

Comment:
This is wrong, `nallocx` returns the excess size of `mallocx` but you are using `rallocx` to allocate (and it doesn't return the excess size but the total size). 

You can either use `sdallocx(ptr, align)` to get the real size allocated for pointer (which I think is more expensive than `nallocx`), or do like #45514 and use `xallocx` to allocate instead of `rallocx` which gives you the total size back.

Reply:
Good point about salloc, I'll change the code to use that.

`xallocx` isn't general, it can "fail to grow" if the reallocation can't be done inplace, see https://github.com/jemalloc/jemalloc/blob/47203d5f422452def4cb29c0b7128cc068031100/test/integration/xallocx.c#L51

Reply:
> `xallocx` isn't general, it can "fail to grow" if the reallocation can't be done inplace

Ineed, I've asked about this in jemalloc's issues earlier today: https://github.com/jemalloc/jemalloc/issues/1059

Ideally what we would like to use is `nallocx` equivalent for `rallocx`, or a `rallocx` variant that returns the excess size. My understanding is that `sdallocx` is correct for all cases, but it can be expensive because it might traverse jemalloc internal data-structures to find the memory.